### PR TITLE
firmware, redirects, tests

### PR DESCRIPTION
**Pull request by mroz22**

### CODE REVIEW

File: packages/suite/src/middlewares/suite/redirectMiddleware.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    const { locks } = api.getState().suite;
+
+    // router is locked, no redirect;
+    if (locks.includes(SUITE.LOCK_TYPE.ROUTER)) {
+        next(action);
+        return action;
+    }
+
+    switch (action.type) {
+        case SUITE.READY:
```
Comment (mroz22) [inline packages/suite/src/middlewares/suite/redirectMiddleware.ts]: note

File: packages/suite/src/actions/suite/suiteActions.ts [hunk 2]
```diff
@@ -153,20 +153,9 @@ export const selectDevice = (device?: Device | TrezorDevice) => async (
...
-        if (device.mode === 'initialize') {
-            await dispatch(goto('onboarding-index'));
-        }
-        // 4. device firmware update required, redirect to "firmware update"
-        else if (device.firmware === 'required') {
-            await dispatch(goto('suite-device-firmware'));
-        }
-    }
-
+    if (locks.includes(SUITE.LOCK_TYPE.DEVICE)) return;
```
Comment (mroz22) [inline packages/suite/src/actions/suite/suiteActions.ts]: ?

File: packages/suite/src/actions/suite/constants/notificationConstants.ts [hunk 3]
```diff
@@ -1,3 +1,7 @@
 export const ADD = '@notification/add';
 export const CLOSE = '@notification/close';
 export const REMOVE = '@notification/remove';
+
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/constants/notificationConstants.ts]: remove unused

File: packages/suite/src/actions/suite/firmwareActions.ts [hunk 4]
```diff
@@ -5,27 +5,23 @@ import { lockUI } from '@suite-actions/suiteActions';
...
 export const firmwareUpdate = () => async (dispatch: Dispatch, getState: GetState) => {
-    // todo remove error notification
-    // dispatch(notificationActions.add({
-    //     variant: "success",
-    //     title: 'hello',
-    // }))
-
-    const { device } = getState().suite;
+    const { device, locks } = getState().suite;
     if (!device || !device.connected || device.type !== 'acquired') {
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/firmwareActions.ts]: device.features

File: packages/suite/src/actions/suite/firmwareActions.ts [hunk 5]
```diff
@@ -59,22 +54,46 @@ export const firmwareUpdate = () => async (dispatch: Dispatch, getState: GetStat
...
+export const enableReducer = (payload: boolean): Action => ({
+    type: FIRMWARE.ENABLE_REDUCER,
+    payload,
+});
 
 interface SetUpdateStatusAction {
-    type: typeof SET_UPDATE_STATUS;
+    type: typeof FIRMWARE.SET_UPDATE_STATUS;
     payload: AnyStatus;
 }
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/firmwareActions.ts]: move types up

File: packages/suite/src/actions/suite/notificationActions.ts [hunk 6]
```diff
@@ -1,19 +1,10 @@
-import * as React from 'react';
-import { FormattedMessage } from 'react-intl';
-import { NotificationProps } from '@trezor/components';
-import { CallbackAction } from '@suite/reducers/suite/notificationReducer';
+// import { FormattedMessage } from 'react-intl';
+// import { NotificationProps } from '@trezor/components';
 import { RequiredKey } from '@suite/types/utils';
+import { NotificationEntry } from '@suite/reducers/suite/notificationReducer';
 import { NOTIFICATION } from './constants';
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/notificationActions.ts:6]: Exclude > Omit

File: packages/suite/src/actions/suite/suiteActions.ts (hunk 2 above)
Comment (szymonlesisz) [reply packages/suite/src/actions/suite/suiteActions.ts]: remove

File: packages/suite/src/actions/suite/suiteActions.ts [hunk 7]
```diff
@@ -386,3 +375,15 @@ export const authorizeDevice = () => async (
...
+
+/**
+ * Wrapper utility action that dispatches:
+ * 1. routerLock(false)
+ * 2. goto(params)
+ *
+ * Useful for exiting apps that operate under locked router.
+ */
+export const exitApp = (routeName: Route['name']) => async (dispatch: Dispatch) => {
+    await dispatch(lockRouter(false));
```
Comment (szymonlesisz) [inline packages/suite/src/actions/suite/suiteActions.ts]: remove await

File: packages/suite/src/middlewares/onboarding/__tests__/onboardingMiddleware.test.ts [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        __esModule: true, // this property makes it work
+        default: () => {},
+    };
+});
+
+jest.mock('@suite-actions/storageActions', () => {
+    return {
+        __esModule: true,
+        loadStorage: () => {
+            return {
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/onboarding/__tests__/onboardingMiddleware.test.ts]: loadStorage doesnt return BLOCKCHAIN.READY

File: packages/suite/src/middlewares/onboarding/onboardingMiddleware.ts [hunk 9]
```diff
@@ -1,33 +1,27 @@
...
-
-    if (
-        action.type === SUITE.APP_CHANGE &&
-        action.payload !== 'onboarding' &&
-        onboarding.reducerEnabled
-    ) {
-        // resetOnboarding also sets enableReducer to false;
-        api.dispatch(resetOnboarding());
-        api.dispatch(lockRouter(false));
+    if (action.type === SUITE.APP_CHANGED) {
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/onboarding/onboardingMiddleware.ts:15]: add more comments whats going on

File: packages/suite/src/middlewares/suite/__tests__/redirectMiddleware.test.ts [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        it('DEVICE.CONNECT mode=initialize', () => {
+            const store = initStore(getInitialState());
+            const goto = jest.spyOn(routerActions, 'goto');
+            store.dispatch({
+                type: DEVICE.CONNECT,
+                payload: getSuiteDevice({ mode: 'initialize' }),
+            });
+
+            expect(goto).toHaveBeenNthCalledWith(1, 'onboarding-index');
+            goto.mockClear();
```
Comment (szymonlesisz) [inline packages/suite/src/middlewares/suite/__tests__/redirectMiddleware.test.ts]: move to "onBeforeEach"

File: packages/suite/src/reducers/onboarding/__tests__/onboardingReducer.test.ts [hunk 11]
```diff
@@ -2,7 +2,7 @@
 /* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
 
 import { UI, DEVICE } from 'trezor-connect';
-import onboardingReducer from '@onboarding-reducers/onboardingReducer';
+import onboardingReducer from '@onboarding-reducers/index';
```
Comment (szymonlesisz) [inline packages/suite/src/reducers/onboarding/__tests__/onboardingReducer.test.ts]: use '@onboarding-reducers'

File: packages/suite/src/reducers/suite/firmwareReducer.ts [hunk 12]
```diff
@@ -20,48 +23,42 @@ export type AnyStatus =
...
-    state: FirmwareUpdateState,
-    draft: Draft<FirmwareUpdateState>,
-    device?: TrezorDevice,
-) => {
-    if (!device || device.type !== 'acquired') {
+const handleSelectDevice = (draft: FirmwareUpdateState, device?: TrezorDevice) => {
+    const prevDevice = draft.device;
+
+    // should not happen but who knows.
+    if (!device || device.type !== 'acquired' || device.mode === 'bootloader') {
```
Comment (szymonlesisz) [inline packages/suite/src/reducers/suite/firmwareReducer.ts]: device.features

File: packages/suite/src/reducers/suite/notificationReducer.ts [hunk 13]
```diff
@@ -18,11 +22,12 @@ export interface NotificationEntry {
     key: string; // autogenerated, used to close notification on click
     id?: string; // programmer provided, might be used to find and close notification programatically
     devicePath?: string; // used to close notifications for device
-    variant: NotificationAddPayload['variant'];
+    variant: 'success' | 'info' | 'error';
     title: React.ReactNode | string;
     message?: React.ReactNode;
     cancelable?: boolean;
     actions?: CallbackAction[];
+    tags?: NotificationTag[];
```
Comment (szymonlesisz) [inline packages/suite/src/reducers/suite/notificationReducer.ts]: remove tag

File: packages/suite/src/views/suite/firmware/index.tsx [hunk 14]
```diff
@@ -72,110 +86,108 @@ const mapDispatchToProps = (dispatch: Dispatch) => ({
...
-    }
+    const hasDevice = () => {
+        return Boolean(device && device.features && device.connected);
+    };
 
-    const isInProgress = () => {
-        if (!firmware || !firmware.status) return null;
-        return ['started', 'downloading', 'installing', 'restarting'].includes(firmware.status);
+    const isInBootloader = () => {
+        return Boolean(device && device.features && device.mode === 'bootloader');
```
Comment (szymonlesisz) [inline packages/suite/src/views/suite/firmware/index.tsx]: use !!(...)

File: packages/suite/src/actions/suite/notificationActions.ts (hunk 6 above)
Comment (mroz22) [reply packages/suite/src/actions/suite/notificationActions.ts:6]: works as expected

**Review by szymonlesisz — APPROVED** [review]

**Pull request merged**